kem/kyber/templates/pkg.templ.go

```diff
@@ -0,0 +1,379 @@
+// +build ignore
+// The previous line (and this one up to the warning below) is removed by the
+// template generator.
+
+// Code generated from pkg.templ.go. DO NOT EDIT.
+
+// {{.Pkg}} implements the IND-CCA2 secure key encapsulation mechanism
+// {{.Name}}.CCAKEM as submitted to round2 of the NIST PQC competition and
+// described in
+//
+// https://pq-crystals.org/dilithium/data/dilithium-specification-round2.pdf
+package {{.Pkg}}
+
+import (
+	"github.com/cloudflare/circl/kem"
+	cpapke "github.com/cloudflare/circl/pke/kyber/{{.Pkg}}"
+
+	"golang.org/x/crypto/sha3"
+
+	"bytes"
+	cryptoRand "crypto/rand"
+	"crypto/subtle"
+	"io"
+)
+
+const (
+	// Size of seed for NewKeyFromSeed
+	KeySeedSize = cpapke.KeySeedSize + 32
+
+	// Size of seed for EncapsulateTo.
+	EncapsulationSeedSize = 32
+
+	// Size of the established shared key.
+	SharedKeySize = 64
+
+	// Size of the encapsulated shared key.
+	CiphertextSize = cpapke.CiphertextSize
+
+	// Size of a packed public key.
+	PublicKeySize = cpapke.PublicKeySize
+
+	// Size of a packed private key.
+	PrivateKeySize = cpapke.PrivateKeySize + cpapke.PublicKeySize + 64
+)
+
+// Type of a {{.Name}}.CCAKEM public key
+type PublicKey struct {
+	pk *cpapke.PublicKey
+
+	hpk [32]byte // H(pk)
+}
+
+// Type of a {{.Name}}.CCAKEM private key
+type PrivateKey struct {
+	sk  *cpapke.PrivateKey
+	pk  *cpapke.PublicKey
+	hpk [32]byte // H(pk)
+	z   [32]byte
+}
+
+// NewKeyFromSeed derives a public/private keypair deterministically
+// from the given seed.
+//
+// Panics if seed is not of length KeySeedSize.
+func NewKeyFromSeed(seed []byte) (*PublicKey, *PrivateKey) {
+	var sk PrivateKey
+	var pk PublicKey
+
+	if len(seed) != KeySeedSize {
+		panic("seed must ne of length KeySeedSize")
+	}
+
+	pk.pk, sk.sk = cpapke.NewKeyFromSeed(seed[:cpapke.KeySeedSize])
+	sk.pk = pk.pk
+	copy(sk.z[:], seed[cpapke.KeySeedSize:])
+
+	// Compute H(pk)
+	var ppk [cpapke.PublicKeySize]byte
+	sk.pk.Pack(ppk[:])
+	h := sha3.New256() // XXX use internal sha3
```

Comment:
Our internal SHA3 clone only contains SHAKE — not SHA3 itself.  I'd like to do some more maintenance to it, but prefer to do that in a separate PR as it will touch some other parts of Circl as well.

Reply:
I see, that's ok.